torch/distributed/_composable_state.py

```diff
@@ -0,0 +1,21 @@
+import torch.nn as nn
+from typing import Any, Callable, cast, Dict, List, Optional
+
+
+class _State:
+    pass
+
+_module_state_mapping: Dict[nn.Module, _State] = {}
+
+def _insert_module_state(module: nn.Module, state: _State) -> None:
+    global _module_state_mapping
+    assert module not in _module_state_mapping
+    _module_state_mapping[module] = state
+
+
+def _get_module_state(module: nn.Module) -> Optional[_State]:
+    global _module_state_mapping
+    if isinstance(module, _State):
+        return cast(_State, module)
+    else:
```

Comment:
just for my understanding, this is the fork in the path for composable vs non-composable scenarios, right?

Reply:
Yes, correct.

Reply:
IIUC: 
- The `if` branch is for non-composable, and the `else` branch is for composable.
- It is not a hard requirement for `DistributedDataParallel` to inherit from some `_DDPState` that inherits from `_State` unless we want to unify the underlying composable/non-composable code like for FSDP.

Reply:
I agree it is not a hard requirement for `DistributedDataParallel` to inherit from `_DDPState` but it won't change anything. If `DistributedDataParallel`  does not unify the code like FSDP but simply branch the code, no issues like FSDP will occur during the shared code path since the code is not unified. The `_DDPState` version can still work for the composable API.

Reply:
The current implementation in `Contract` is having a State per module per API. This PR will change it to per module State shared by all APIs, is my understanding correct?
cc @mrshenli